### PR TITLE
Add a custom macro to allow generation of a section index page

### LINKED ISSUE #822: create single linkable page or section with overview of data sources
From @Bengoldacre: https://bennettoxford.slack.com/archives/C31D62X5X/p1656065228481669

> the problem is the things you'd naturally like to link to are folders in the nav with rotating arrows, not pages ("Data sources" and maybe "Externally linked data") 

So we would like an additional section that links to all available data sources available on each backend. An overview of _all_ available data sources across _all_ backends on a single page doesn't quite fit with the existing structure of backend-specific pages (eg [for TPP](https://docs.opensafely.org/dataset-systmone/)) but that's ok -- it can go on the [overview](https://docs.opensafely.org/dataset-intro/) page as a nested list like this:

* OpenSAFELY-TPP
  * SystmOne primary care
  * SUS hospital admissions
  * SUS outpatients
  * death register
  * ...
* OpenSAFELY-EMIS
  * EMIS primary care
  * ?
* (other future operational backends)
  * data

### PR DESCRIPTION
**Pull request by rebkwok**
Fixes #822 

This allows us to generate a list of nested links on an index page for any given section of the navigation.

It involves using the [section indexes](https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#section-index-pages) feature, which links an index page within a section subfolder to a section in the nav, and using the [macros plugin](https://mkdocs-macros-plugin.readthedocs.io/) which makes the global navigation available, plus a custom macro to generate the HTML to display. 

Note that it doesn't generate the different list structure exactly as specified in #822 - it just duplicated the nav structure for the section into a list on the index page.

### CODE REVIEW

**Comment by StevenMaude** [comment]
This looks fine to me, after probably a lot of wrangling on @rebkwok's part.

This will change the URLs for the data sources pages affected. That's not necessarily a bad thing, and it would be better if the URLs had more structure (see #717).

But we might want to redirect the existing URLs so we don't break anyone's bookmarks, and for maintaining search engine rankings.

It's configured [in this repository](https://github.com/opensafely/documentation/blob/22a2701ae12364ed8500a693de60081d9c8a6df1/render.yaml) now.

**Comment by rebkwok** [comment]
@StevenMaude I tried adding the redirects to the render.yaml, but they're not working - e.g. https://opensafely-documentation-pr-827.onrender.com/dataset-emis 404s.  Is there some precendence that I'm breaking there?

Weirdly https://opensafely-documentation-pr-827.onrender.com/dataset-intro 404s, but https://opensafely-documentation-pr-827.onrender.com/dataset-intro/ does not (but it's not redirecting to /datasets/dataset-intro)

File: main.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+def define_env(env):
+    "Hook function"
+
+    @env.macro
+    def build_toc(nav, page):
+        """
+        Build a nested list of page links from a navigation section index, to be inserted into the index page iteself
```
Comment (StevenMaude) [inline main.py:7]:
```suggestion
        Build a nested list of page links from a navigation section index, to be inserted into the index page itself
```

File: main.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+
+def make_link(item):
+    """
+    Create the html list links for a nav item
+    `item` may be a Section or a Page
+    """
+    if item.is_page:
+        # Note that we prepend / so the URL extracted from the navigation is relative to the
+        # root, and not to the location of this index page.  This means we can deal with nested # nav links from any point in the document structure
```
[the remark this answers is not in the document]
Comment (rebkwok) [reply main.py:23]: huh, I didn't notice because my IDE wraps lines automatically (hence why I've put a random `#` in the middle of the line)

**Pull request merged**